### PR TITLE
Raduom/marconi initiative

### PR DESCRIPTION
**Pull request by raduom**
Basically put down some thoughts on why Marconi makes sense.

PR: https://input-output.atlassian.net/jira/software/c/projects/PLT/boards/788?modal=detail&selectedIssue=PLT-503

This is an ADR.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:5]:
```suggestion
Plutus off-chain code oftentimes needs to have access to data thats has been already stored on-chain. The Chain Index project is a solution to deliver access to this data by indexing portions of the blockchain. However, after release, few shortcomings were identified which prompted the development of an indexing solution based on different architectural and functional constraints.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:15]: To me this sounds just like a PAB bug which can be solved without a re-architecture. Like point B above.

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:17]: Sounds a bit like point B but `lack of non-functional requirements` is importan on its own. Perhaps merge with point E below?

e.g.

```
B. The monolithic architecture lacks the features necessary to customise the indexed set of data, the Chain Index provides only all-or-nothing indexing. Without changing the code, it impossible to turn off indexing of data that is not required by our customers. While this can be addressed in principle, the monolithic architecture makes it an uphill battle.

E. The lack of non-functional requirements resulted in software that uses an unreasonable amount of resources and results in slow synchronisation speeds. Combined with the lack of a specification this makes the testing feel ad-hoc and like an afterthought.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:43]:
```suggestion
Any indexing solution has to deal with blockchain rollbacks, which can happen as the network achieves consensus. We introduce the following basic requirements: it needs to deal with rollbacks as elegantly as possible and provide a way to compromise between memory, disk and CPU usage.
```

Expanding a bit just because you haven't mentioned rollbacks before.

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:49]: What is "them"? You save the persisted state to disk past the security parameter right? not the blocks themselves.

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:53]:
```suggestion
  1. Managing rollbacks is very simple and fast. We drop the events that were rolled back. (no need to undo the application of blocks on the state stored on disk, which would be necessary if we were to store everything on disk as fast as possible).
```

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:55]: > thus sacrificing chain integrity for RAM

I think this point is very important but "sacrificing chain integrity" with no extra detail sounds a bit scary. Can you expand?

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:57]:
```suggestion
  3. In case of a restart recovery is very simple. We only store fully confirmed transactions so there is nothing to do other than loading the state persisted on disk and following the blockchain from the corresponding chain-point.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:61]: > can waste some memory

I don't think it's necessarily a "waste".

```suggestion
  1. We must keep K events in memory, which (depending on how large events are) can waste some memory. Our educated guess is that this is a reasonable compromise, but depending on how large events can get that may not be the case for every use case.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:71]: We definitely need a specification for this API.

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [inline doc/adr/0004-marconi-initiative.rst:79]: I am not sure what you mean by this.

**Review by andreabedini — COMMENTED** [review]
This is good. I allowed myself to reword some parts but feel free to take what you like and discard the rest. I commented on a couple of points that are unclear to me.

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:5]: Suggestion: 
"to have access to data that has already been stored..." 
"However, after release, a few shortcomings..."

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:11]: As a non-subject matter expert reader in this area, I don't know the meaning of the term "free algebras," but I imagine it is possible that the target audience may be expected to be familiar with this term.

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:15]: Suggestion to add a couple of commas: 
Line 13: "While this can be addressed, the monolithic..." 
Line 15: "...that assumes that there is only one index running, we ran into..."

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [inline doc/adr/0004-marconi-initiative.rst:21]: The Chain Index was meant to be a software application that supports the execution of smart contracts. And, in that, it succeeded. However, we found that our customers would rather have a library of functionality that they can customize to do the following:
* to build their own indexers, 
* to work only with the data that they care about for their application, 
* to use whatever storage engine they prefer, and 
* to support only the queries that they need to support.

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:53]: Suggestion: Use a capital "N" --> (No need to undo the application... )

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:55]: Suggestion: reverse order of "only" and "store" --> 
"...so they can choose to store only 10 events in memory..."

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:57]: Suggest adding a comma: 
"In case of a restart, recovery is very simple."

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:61]: Suggestion: 
"Our educated guess is that this is a reasonable compromise; however, depending on how large events can get, that may not be..."

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [inline doc/adr/0004-marconi-initiative.rst:91]: Suggest adding a comma: 
"...for a set of event types, we need a way..."

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [inline doc/adr/0004-marconi-initiative.rst:99]: I'm slightly unclear on the meaning. Can you clarify a little bit? Is this saying that Marconi is both a streaming solution and an indexer? I'm not sure what the word "they" is referring to --> "they are both a streaming solution and an indexer" 

I like this document! Very informative.

**Review by joseph-fajen — APPROVED** [review]
Radu, I left some comments. I found this to be a very informative document!

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:5]:
```suggestion
Plutus off-chain code oftentimes needs access to indexed portions of the blockchain. The plutus-chain-index project is the initial solution meant to deliver access to this kind of data. However, after release, a couple of shortcomings were identified which prompted the development of an indexing solution that is based on a different set of architectural and functional constraints.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:7]:
```suggestion
A lot of the shortcomings are connected to the exploratory type of development that we used to deliver the plutus-chain-index which was prompted by the lack of a clear specification and a lack of concern for non-functional and quality assurance requirements. The top-down design resulted in a monolithic and fairly complex architecture which made the code difficult to reuse, compose and understand.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:11]: I understand the point, but I don't think I'm in agreement. The use of an effect system in `plutus-chain-index` was to, like you said, have two implementations: one for the emulator and one for production use. While I agree with the point that "there would be a lot of production code that would not be tested", saying that the effect system "makes the code fairly complex and difficult to understand" is too opinionated. I believe we could have a similar problem (havingcomplex code) even if we tried to reproduce the same behavior with `mtl` instead of `freer-simple` (there's even more boilerplate with `mtl` IMO).

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [reply doc/adr/0004-marconi-initiative.rst:15]: To continue @andreabedini's point, this is a PAB architectural bug which is addresses here: #550 . Once the PR is merged I'll reference it to this ADR.

Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:15]:
```suggestion
  B. The monolithic architecture lacks the features necessary to customise the indexed set of data, the Chain Index provides only all-or-nothing indexing. While this can be addressed the monolithic architecture makes it an uphill battle. 
  
  C. Also due to the monolithic architecture that assumes that there is only one index running we ran into trouble when we made the Plutus Application Backend collect and index information requested by smart contracts. Now we have two components that index information from the blockchain, but they are not synchronised. For example, querying the Chain Index about a transaction, from which the PAB confirmed it's presence in a Cardano block, often results in no data returned, since the Chain Index indexes data slower than the PAB.
```

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:21]: Well. "Succeeded" is probably not the right term. I would say that functional requirement were met, but non-functional requirements were not adequately thought about, which resulted in a bad experience for users.

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:31]: I -> We?

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:45]: We need to reconfirm this fact. I read an important detail about the K parameter. I always though that it meant that once a block is K blocks deep in the blockchain, then it is immutable. However, that is not the case. The real formula to determine how deep a block should be in order for it to be immutable is something like (I don't remember it exactly): 2 * K / activeSlotCoef. I'll need to research this more.

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [reply doc/adr/0004-marconi-initiative.rst:71]: There's an ADR coming for this API in the next sprint probably. Right @raduom ?

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:81]: Link to the other ADR once available.

**Comment by koslambrou** [comment]
@raduom This ADR will probably get a lot of feedback/comments. I would suggest:

* to keep this PR open for some sufficient amount of time in order to allow everyone time to put in their comments
* mark it as draft in order to not merge it yet

What do you think?

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:99]: This section should (maybe? probably?) be expanded.

Comment (koslambrou) [inline doc/adr/0004-marconi-initiative.rst:99]: Links to Oura and Scrolls would be helpful.

File: doc/adr/0004-marconi-initiative.rst
Comment (joseph-fajen) [reply doc/adr/0004-marconi-initiative.rst:71]: Question about line 69: "...the code for both of them has to be provided by the user." Can you clarify what "both of them" is referring to? Is it the queries and the type variable? Or the results?

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:53]: You never rollback things that are on-disk. That is why the indexers are fairly simple.

**Comment by raduom** [comment]
> 

Done.

**Comment by raduom** [comment]
> * mark it as draft in order to not merge it yet

I have marked it as draft.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:99]: I have this slight problem which is the slot number changes which may add some updates to this document and I also don't know very well how oura/scrolls work.

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [reply doc/adr/0004-marconi-initiative.rst:53]: Oh! Then with my current understand, I would heavily object to this decision. Here's my perspective:

If you never rollback things that are on-disk, then it makes no sense to give the user the option of defining how many blocks are kept in memory. You really don't want to affect the integrity of the on-disk data.

IF we want to give the user the option to specify the amount of blocks in memory (like 100 for example), then we definitely **need** to handle rollbacks on data that is on disk. Customizability is one of the main key points of Marconi. Doesn't made sense to break chain data integrity we need to  rollback, let's say, 500 blocks deep. Plus, I though the indexer (hysterical-screams) was meant to make the process of implementing rollbacks simple to handle between data that is on disk and data that is on memory.

File: doc/adr/0004-marconi-initiative.rst
Comment (koslambrou) [reply doc/adr/0004-marconi-initiative.rst:99]: Then, we can wait for Joseph's work which compares the different tools.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:15]: @andreabedini I am not really convinced about C not being an architectural bug. You can sort of solve it if your indexing solution fits into one machine (and note, that it may need to fit into a developer's machine, not necessarily a server), but if it does not, then you will need some kind of distribution anyway, in which case life gets complicated for chain-index.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:17]: B is about the inability of splitting up indexing into several components that can be optionally turned off, whereas E (which is the point I suppose you mean) is about the lack of a testing strategy.

File: doc/adr/0004-marconi-initiative.rst
Comment (andreabedini) [reply doc/adr/0004-marconi-initiative.rst:15]: I think we can move the discussion to #550.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:71]: > We definitely need a specification for this API

Is the type signature of the required functions what you had in mind?

Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:71]: > Question about line 69: "...the code for both of them has to be provided by the user." Can you clarify what "both of them" is referring to? Is it the queries and the type variable? Or the results?

I clarified (I hope) the meaning.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:79]: I rewrote it. Hopefully it's a little more clear now.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:63]: I removed both and replaced the word `query` with `scan`.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:99]: I need to rewrite this, but I think I need a little bit of time to look over the competing solutions to make sure I am not assuming too many things about them.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:5]: I renamed all the `Chain Index` strings to `plutus-chain-index`. I think it looks better.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:11]: I am not the only person that said that our code is "fairly complex and difficult to understand". At least one of our clients pointed that to us in a previous occasion. I was referring to the type-level programming here, which is not so prevalent in a MTL based solution. It was not about boilerplate as much as complexity.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:21]: What I meant here is that it was usable. And it still is usable. It can be faster, more modular, etc. But it works, which is a great achievement given the lack of specification and complexity of the problem, which is something I wanted to give props for.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:45]: As long as there is a formula, there is a K.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst:57]: Updated no. 3.

File: doc/adr/0004-marconi-initiative.rst
Comment (raduom) [reply doc/adr/0004-marconi-initiative.rst]: > If you never rollback things that are on-disk, then it makes no sense to give the user the option of defining how many blocks are kept in memory. You really don't want to affect the integrity of the on-disk data.

I would not mind giving this option to developers working on machines that have very little ram, however for production-level indexers it would be pretty bad.

The way we did it before is that we hid this option under a 'development options' header. We could have a development-mode run configuration which allows setting K < 2160 and a production-mode run configuration that only allows for K > 2160.

As a different alternative, we can always complicate the indexers to handle on-disk rollbacks and provide that as an option, in case users don't mind a bit more complexity.

**Review by joseph-fajen — APPROVED** [review]
Looks like some helpful clarifications here.

**Comment by koslambrou** [comment]
> I do think we need further discussion on handling rollbacks greater than K

@asutherlandus After our discussion with Mamba, it seems like there's no need to handle rollbacks greater than K, because there won't be any. Once the block is K (2160 on the current mainnet) blocks deep (with an average of 1 block every 20s), the block should be immutable. That's what I understood from the conversation.

**Review by andreabedini — APPROVED** [review]
LGTM

**Pull request merged**
